### PR TITLE
Add projection for sparse Jacobian matrices

### PR DESCRIPTION
**Pull request by fantaosha**
## Motivation and Context
Update `LieGroup.project()` for sparse jacobian matrices.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: theseus/geometry/tests/test_se2.py [hunk 1]
```diff
@@ -128,21 +128,42 @@ def test_projection():
...
 
             # Test SE2.transform_to
-            def transform_to_sum(g, p):
-                return th.SE2(data=g).transform_to(p).data.sum(dim=0)
+            def transform_to_func(g, p):
+                return th.SE2(data=g).transform_to(p).data
 
-            jac = torch.autograd.functional.jacobian(
-                transform_to_sum, (se2.data, point.data)
+            jac_raw = torch.autograd.functional.jacobian(
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/tests/test_se2.py:139]: I will check this now.

[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/tests/test_se2.py]: Refactored and added `check_projection()` [here](https://github.com/facebookresearch/theseus/blob/ac7ac0a93229950ca6e53e7402da5dd3b4a499ee/theseus/geometry/tests/common.py#L61).

File: theseus/geometry/tests/common.py [hunk 2]
```diff
@@ -56,3 +56,39 @@ def check_adjoint(group, tangent_vector):
         group_matrix @ group.hat(tangent_vector) @ group.inverse().to_matrix()
     )
     assert torch.allclose(tangent_left.squeeze(2), tangent_right, atol=EPS)
+
+
+def check_projection(Group, Point, Func, batch_size, generator):
+    group = Group.rand(batch_size, generator=generator, dtype=torch.float64)
+    point = Point.rand(batch_size, generator=generator, dtype=torch.float64)
+    aux_id = torch.arange(batch_size)
```
Comment (luisenp) [inline theseus/geometry/tests/common.py:64]: Move `aux_id` to line 80 so it's easier to read the overall logic.

File: theseus/geometry/tests/common.py [hunk 3]
```diff
@@ -56,3 +56,39 @@ def check_adjoint(group, tangent_vector):
         group_matrix @ group.hat(tangent_vector) @ group.inverse().to_matrix()
     )
     assert torch.allclose(tangent_left.squeeze(2), tangent_right, atol=EPS)
+
+
+def check_projection(Group, Point, Func, batch_size, generator):
```
Comment (luisenp) [inline theseus/geometry/tests/common.py:61]: Add a comment reminder that `Func` returns the result of some group operation (e.g., transform_to, inverse, etc.) cost and also populates jacobians in the list passed as the third argument.

File: theseus/geometry/tests/test_se2.py [hunk 4]
```diff
@@ -127,43 +128,5 @@ def test_projection():
...
-            # Check sparse jacobian matrices
-            actual = [
-                se2.project(jac_raw[0][aux_id, :, aux_id, :], is_sparse=True),
-                point.project(jac_raw[1][aux_id, :, aux_id, :], is_sparse=True),
-            ]
-
-            expected = jac
-            assert torch.allclose(actual[0], expected[0])
-            assert torch.allclose(actual[1], expected[1])
+            check_projection(th.SE2, th.Point2, th.SE2.transform_to, batch_size, rng)
```
Comment (luisenp) [inline theseus/geometry/tests/test_se2.py:132]: Is it worth it to check with other group functions as well? (e.g., a loop over `inverse`, `compose`, etc.)

**Review by luisenp — CHANGES_REQUESTED** [review]
The `check_projection` code looks good! Left some smaller comments to discuss/address.

File: theseus/geometry/tests/common.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/tests/common.py:63]: Updated. Let me know if it is ready to merge.

File: theseus/geometry/tests/test_se2.py (hunk 4 above)
Comment (luisenp) [reply theseus/geometry/tests/test_se2.py:132]: Yes, we do have Jacobians for all of these.

Comment (fantaosha) [reply theseus/geometry/tests/test_se2.py]: Added `check_projection_for_compose` and `check_projection_for_inverse` [here](https://github.com/facebookresearch/theseus/blob/41c3e90e283065d15804bd2d439d4d8a02ab6227/theseus/geometry/tests/common.py#L103).

File: theseus/geometry/tests/common.py [hunk 5]
```diff
@@ -56,3 +56,161 @@ def check_adjoint(group, tangent_vector):
...
+):
+    group = Group.rand(batch_size, generator=generator, dtype=torch.float64)
+    point = Point.rand(batch_size, generator=generator, dtype=torch.float64)
+
+    def func(g, p):
+        return Func(Group(data=g), p).data
+
+    jac_raw = torch.autograd.functional.jacobian(func, (group.data, point.data))
+    jac = []
+    _ = Func(group, point, jac)
```
Comment (mhmukadam) [inline theseus/geometry/tests/common.py:74]: Why not also assert on the returned value from Func (in addition to jacobians) or is this tested elsewhere separately.

File: theseus/geometry/tests/common.py [hunk 6]
```diff
@@ -56,3 +56,161 @@ def check_adjoint(group, tangent_vector):
...
+
+    # Check dense jacobian matrices
```
Comment (mhmukadam) [inline theseus/geometry/tests/common.py:76]: Are you also testing sparse jacobians?

File: theseus/geometry/tests/common.py (hunk 5 above)
Comment (fantaosha) [reply theseus/geometry/tests/common.py:74]: Will add this.

File: theseus/geometry/tests/common.py (hunk 6 above)
Comment (fantaosha) [reply theseus/geometry/tests/common.py:76]: Yes, it is checked [here](https://github.com/facebookresearch/theseus/blob/41c3e90e283065d15804bd2d439d4d8a02ab6227/theseus/geometry/tests/common.py#L92).

Comment (mhmukadam) [reply theseus/geometry/tests/common.py:76]: Ah thanks for the pointer, missed it.

File: theseus/geometry/tests/common.py (hunk 5 above)
Comment (fantaosha) [reply theseus/geometry/tests/common.py]: > Why not also assert on the returned value from Func (in addition to jacobians) or is this tested elsewhere separately.

Added.

**Review by mhmukadam — APPROVED** [review]

**Review by luisenp — APPROVED** [review]

**Pull request merged**